### PR TITLE
Expose components via functions with params merged with defaults

### PR DESCRIPTION
**Pull request by metalmatze**
Going forward the components exposed by this project are going to be functions.
That means that a bit of refactoring will be necessary to go to a next version.
It's best to look at our examples in this PR to see what needs changing, it shouldn't be much, as the config object is also used as a parameter object now.

With this new approach we hope to better scope/communicate the out-of-the-box configurable things compared to the ones that need to be overwritten when encountered. Meaning, everything that's not configurable is to be merged on top of the objects generated by the functions.

### CODE REVIEW

**Review by kakkoyun — COMMENTED** [review]
Looks great 🥇 Only small nits

File: examples/all/manifests/thanos-query-deployment.yaml [hunk 1]
```diff
@@ -71,5 +71,6 @@ spec:
             port: 9090
             scheme: HTTP
           periodSeconds: 5
+        resources: {}
```
[the remark this answers is not in the document]
Comment (metalmatze) [reply examples/all/manifests/thanos-query-deployment.yaml:74]: It'll be ignored like this.

File: jsonnet/kube-thanos/kube-thanos-compact.libsonnet [hunk 2]
```diff
@@ -1,27 +1,46 @@
...
-    podLabelSelector:: {
-      [labelName]: tc.config.commonLabels[labelName]
-      for labelName in std.objectFields(tc.config.commonLabels)
-      if !std.setMember(labelName, ['app.kubernetes.io/version'])
-    },
-  },
+  // Combine the defaults and the passed params to make the component's config.
+  config:: defaults + params,
+  // Safety checks for combined config of defaults and params
+  assert std.isNumber(tc.config.replicas) && tc.config.replicas >= 0 : 'thanos compact replicas has to be number >= 0',
```
[the remark this answers is not in the document]
Comment (metalmatze) [reply jsonnet/kube-thanos/kube-thanos-compact.libsonnet:39]: Right. We should make it 0 or 1. I still can be 0. :)

**Review by kakkoyun — APPROVED** [review]
👍

**Pull request merged**